apps/wing-console/console/ui/src/layout/default-layout.tsx

```diff
@@ -239,6 +245,40 @@ export const DefaultLayout = ({
     ],
   );
 
+  const [metadataInstances, setMetadataInstances] = useState<
+    Record<string, MetadataInstance>
+  >({});
+
+  useEffect(() => {
+    const currentNode = selectedItems[0];
+    if (!currentNode) {
+      return;
+    }
+
+    setMetadataInstances((instances) => {
+      instances[currentNode] = {
+        id: currentNode,
+        data: metadata.data,
+      };
+      return instances;
+    });
+  }, [metadata.data, selectedItems]);
+
+  useEffect(() => {
+    const node = items[0];
+    if (!node) {
+      return;
+    }
+
+    const newInstances = { ...metadataInstances };
+    for (const nodeId of Object.keys(metadataInstances)) {
+      if (!hasChild(node, nodeId)) {
+        delete newInstances[nodeId];
+      }
+    }
+    setMetadataInstances(newInstances);
+  }, [items]);
```

Comment:
You should pass a callback inside `setMetadataInstances` or otherwise it won't correctly work since the effect depends on `metadataInstances`

Reply:
you're right!